src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -2810,7 +2810,375 @@ tx_gateway_frontend::reabort_tm_tx(
     co_return tx;
 }
 
-// get_tx must be called under stm->get_tx_lock
+ss::future<checked<tm_transaction, tx_errc>> tx_gateway_frontend::bump_etag(
+  model::term_id term,
+  ss::shared_ptr<cluster::tm_stm> stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout) {
+    checked<tm_transaction, tx_errc> r1(tx);
+    if (tx.status == tm_transaction::tx_status::prepared) {
+        r1 = co_await recommit_tm_tx(stm, term, tx, timeout);
+    } else if (tx.status == tm_transaction::tx_status::aborting) {
+        r1 = co_await reabort_tm_tx(stm, term, tx, timeout);
+    } else if (tx.status == tm_transaction::tx_status::killed) {
+        r1 = co_await reabort_tm_tx(stm, term, tx, timeout);
+    }
+    if (!r1.has_value()) {
+        vlog(
+          txlog.warn,
+          "got error {} on rolling previous tx.id={} with status={}",
+          r1.error(),
+          tx.id,
+          tx.status);
+        // until any decision is made it's ok to ask user retry
+        co_return tx_errc::not_coordinator;
+    }
+
+    if (tx.etag != term) {
+        tx.etag = term;
+        auto r2 = co_await stm->update_tx(tx, term);
+        if (!r2) {
+            vlog(
+              txlog.info,
+              "got {} on bumping etag on reading {}",
+              r1.error(),
+              tx.id);
+            co_return tx_errc::not_coordinator;
+        }
+        co_return r2.value();
+    }
+
+    co_return tx;
+}
+
+ss::future<checked<tm_transaction, tx_errc>> tx_gateway_frontend::forget_tx(
+  model::term_id term,
+  ss::shared_ptr<cluster::tm_stm> stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout) {
+    checked<tm_transaction, tx_errc> r1(tx);
+    if (tx.status == tm_transaction::tx_status::prepared) {
+        r1 = co_await recommit_tm_tx(stm, term, tx, timeout);
+    } else if (tx.status == tm_transaction::tx_status::aborting) {
+        r1 = co_await reabort_tm_tx(stm, term, tx, timeout);
+    }
+
+    if (!r1.has_value()) {
+        vlog(
+          txlog.warn,
+          "got error {} on rolling previous tx.id={} with status={}",
+          r1.error(),
+          tx.id,
+          tx.status);
+        // until any decision is made it's ok to ask user retry
+        co_return tx_errc::not_coordinator;
+    }
+
+    // TODO: check result
```

Comment:
while I was writing this code I noticed that expire_tx doesn't return error code and it may cause bugs in the already existing branches; fixing expire_tx is orthogonal to current commit so added a todo and fixed in the follow up commit